pkg/apis/clusteroperator/types.go

```diff
@@ -225,26 +250,34 @@ type MachineSetSpec struct {
 	NodeLabels map[string]string
 }
 
+// MachineSetSpec is the specification for a MachineSet
+type MachineSetSpec struct {
+	// MachineSetConfig is the configuration for the MachineSet
+	MachineSetConfig
+}
+
 // MachineSetHardwareSpec specifies the hardware for a MachineSet
 type MachineSetHardwareSpec struct {
-	// +optional
 	// AWS specifies the hardware spec for an AWS MachineSet
+	// +optional
 	AWS *MachineSetAWSHardwareSpec
 }
 
 // MachineSetAWSHardwareSpec specifies AWS hardware for a MachineSet
 type MachineSetAWSHardwareSpec struct {
 	// InstanceType is the type of instance to use for machines in this MachineSet
+	// +optional
 	InstanceType string
```

Comment:
We are going to need to validate in the API (1) that the Cluster and the ClusterMachineSet use the same type of hardware spec and (2) that each field in the hardware spec that **needs** a value has one in either the Cluster or the ClusterMachineSet.

This can be done in a follow-on if you write up an issue for it.

Reply:
https://github.com/openshift/cluster-operator/issues/33